### PR TITLE
Make shared workers available in worker threads only

### PR DESCRIPTION
**Pull request by dnlup**
Allow the use of shared workers only when worker threads are enabled.

Fixes: #2705 

Note:

I am going to try to address https://github.com/avajs/ava/issues/2705#issuecomment-792329194 also.

### CODE REVIEW

File: lib/plugin-support/shared-worker-loader.js [hunk 1]
```diff
@@ -15,8 +15,17 @@ class TestWorker {
 	constructor(id, file) {
 		this.id = id;
 		this.file = file;
+		this.sharingMap = new Map();
+		this.transferringMap = new Map();
 	}
 
+	sharing(...buffers) {
```
Comment (dnlup) [inline lib/plugin-support/shared-worker-loader.js:22]: @novemberborn Not sure if this is the API you have in mind.

Comment (novemberborn) [reply lib/plugin-support/shared-worker-loader.js:22]: I think it has to return something new where you can share more, or publish.

Comment (dnlup) [reply lib/plugin-support/shared-worker-loader.js:22]: So like a new `TestWorker` initialized with the shared buffer(s)?

File: lib/plugin-support/shared-worker-loader.js [hunk 2]
```diff
@@ -112,6 +121,14 @@ const nextMessageId = () => `${messageIdPrefix}/${++messageCounter}`;
 
 function publishMessage(testWorker, data, replyTo) {
 	const id = nextMessageId();
+	if (testWorker.sharingMap.size) {
```
Comment (dnlup) [inline lib/plugin-support/shared-worker-loader.js:124]: I am assuming `data` is an object and overriding the previously shared buffers in the payload. I am not sure this works when finally serializing `data` with `v8`.

File: lib/plugin-support/shared-worker-loader.js (hunk 1 above)
Comment (dnlup) [reply lib/plugin-support/shared-worker-loader.js:22]: Mmh, maybe a `Message` class

**Comment by novemberborn** [comment]
@dnlup I like the feature check!

Before you can transfer or use shared memory though you'll have to rework the message-passing logic currently in place for shared workers. Instead of all the forwarding and serializing it should be possible to create direct message channels between each shared worker and the plugin code in the test worker.

That might be enough for one PR.

---

Regarding the interface for transferring and sharing, perhaps we should bike-shed the TypeScript definition instead of diving into the implementation. It may be easier to work out the possibilities that way.

**Comment by dnlup** [comment]
> Instead of all the forwarding and serializing it should be possible to create direct message channels between each shared worker and the plugin code in the test worker.

Ok, I understand. I have to get a better understanding of the architecture of the shared worker then, thanks.

> That might be enough for one PR.

You mean this PR, right?

> Regarding the interface for transferring and sharing, perhaps we should bike-shed the TypeScript definition instead of diving into the implementation. It may be easier to work out the possibilities that way.

That makes sense. I have to admit I am struggling to find a way to implement chaining without writing inefficient logic (like tracking what is shared and always look up for it when publishing a message). I think I need some time to figure that out. Maybe. like you said, working on the TS definitions first will help.

**Comment by dnlup** [comment]
Before I try to rework the communication, I'll revert the latest commit implementing `sharing` as a concept.

**Comment by dnlup** [comment]
It's still not clean. I was able to set up a dedicated channel between the shared worker and the test worker(s). I only have to figure out one failure that I don't understand atm (https://github.com/avajs/ava/pull/2706/checks?check_run_id=2165465306#step:7:4362). It does fail locally on my machine, it seems to pass on some others in the CI. There's some async flow I must be missing because the test in the fixtures fails with `Promise returned by test never resolved` even if it looks like all the asserts are run and passed.  In the meantime, if there's something in the API that could be removed or simplified let me know.

**Review by novemberborn — COMMENTED** [review]
Nice!

Not immediately sure why the tests are failing on Windows. I should have a proper go at tracing the messages but I'm not in the right mind for that kind of task at the moment 😉

File: lib/fork.js [hunk 3]
```diff
@@ -16,13 +16,17 @@ const AVA_PATH = path.resolve(__dirname, '..');
...
 class SharedWorkerChannel extends Emittery {
-	constructor({channelId, filename, initialData}, sendToFork) {
+	constructor({channelId, filename, initialData}, worker) {
 		super();
 
 		this.id = channelId;
 		this.filename = filename;
 		this.initialData = initialData;
-		this.sendToFork = sendToFork;
+		this.worker = worker;
```
[the remark this answers is not in the document]
Comment (dnlup) [reply lib/fork.js:25]: Yes, it's the test worker. I definitely need to improve naming across all my changes. Thanks.

File: lib/plugin-support/shared-worker-loader.js [hunk 4]
```diff
@@ -12,9 +12,10 @@ const events = new EventEmitter().setMaxListeners(0);
 const activeTestWorkers = new Map();
 
 class TestWorker {
-	constructor(id, file) {
+	constructor(id, file, port) {
 		this.id = id;
 		this.file = file;
+		this.port = port;
```
[the remark this answers is not in the document]
Comment (dnlup) [reply lib/plugin-support/shared-worker-loader.js:18]: Ok, how do you feel about using `Symbol` instead?

File: lib/plugin-support/shared-worker-loader.js [hunk 5]
```diff
@@ -129,12 +131,18 @@ function publishMessage(testWorker, data, replyTo) {
 }
 
 function broadcastMessage(data) {
+	// TODO: broadcasting can be done directly here
+	// sice we have a direct channel with each testWorker.
+	// Not sure what to return though.
```
[the remark this answers is not in the document]
Comment (dnlup) [reply lib/plugin-support/shared-worker-loader.js:136]: It indeed seems to work by returning the same value.

File: lib/plugin-support/shared-worker-loader.js [hunk 6]
```diff
@@ -129,12 +131,18 @@ function publishMessage(testWorker, data, replyTo) {
...
-		type: 'broadcast',
-		messageId: id,
-		serializedData: [...v8.serialize(data)]
-	});
+	const serializedData = [...v8.serialize(data)];
+	for (const trackedWorker of activeTestWorkers.values()) {
+		trackedWorker.instance.port.postMessage({
+			type: 'broadcast',
+			messageId: id,
+			serializedData
```
[the remark this answers is not in the document]
Comment (dnlup) [reply lib/plugin-support/shared-worker-loader.js:143]: Definitely, I'll remove it.

File: lib/plugin-support/shared-workers.js [hunk 7]
```diff
@@ -16,34 +15,39 @@ const waitForAvailable = async worker => {
...
 		workerData: {
 			filename,
 			id,
 			initialData
 		}
 	});
-	worker.setMaxListeners(0);
+	sharedWorker.setMaxListeners(0);
+
+	// Await setupChannel(sharedWorker, testWorker, filename);
```
[the remark this answers is not in the document]
Comment (dnlup) [reply lib/plugin-support/shared-workers.js:37]: If we used it should have to be `async` (there's some message passing to wait for a port to be set up), but we don't need it 😄 . The `MessageChannel` has an internal queue for the messages, we don't have to wait for anything to use it.

File: lib/plugin-support/shared-worker-loader.js [hunk 8]
```diff
@@ -182,14 +190,23 @@ loadFactory(workerData.filename).then(factory => {
...
 				if (message.type === 'register-test-worker') {
-					const {id, file} = message;
-					const instance = new TestWorker(id, file);
+					const {id, file, port} = message;
+					const instance = new TestWorker(id, file, port);
 
 					activeTestWorkers.set(id, {instance, teardownFns: new Set()});
 
 					produceTestWorker(instance);
+					setupPort(instance.port);
```
[the remark this answers is not in the document]
Comment (dnlup) [reply lib/plugin-support/shared-worker-loader.js:209]: > Nice!
> 
> Not immediately sure why the tests are failing on Windows. I should have a proper go at tracing the messages but I'm not in the right mind for that kind of task at the moment

No problems, thank you for reviewing even at this unfinished state. It fails on my machine too and I am using Ubuntu, so there's something I have to check out for sure. I'll work on your suggestions in the meantime 😉

File: lib/plugin-support/shared-worker-loader.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (novemberborn) [reply lib/plugin-support/shared-worker-loader.js:18]: Yea that could work. As long as it's clear that the message port is not part of the shared worker API.

**Comment by dnlup** [comment]
I really don't know yet why tests are passing here, but not locally.

The test that fails is `worker-startup-crash` with this error:

```bash
Error: Command failed with exit code 1: /home/dnlup/.volta/tools/image/node/14.16.0/bin/node --require /home/dnlup/Dev/dnlup/ava/test/helpers/simulate-tty.js /home/dnlup/Dev/dnlup/ava/cli.js
  ⚠ Experiments are enabled. These are unsupported and may change or be removed at any time.

  ✖ Error in shared worker

  Error: 🙈

  › crash (_worker.js:10:9)

  ✖ shared workers crash Promise returned by test never resolved
  ─

  shared workers crash

  Error: Promise returned by test never resolved
```

**Comment by novemberborn** [comment]
@dnlup so I didn't end up looking into that specific error, but at the overall approach. I think we got stuck in making and discussing local changes. Now that shared workers only work in threads we can actually allocate the message ports in the test worker and forward them to the shared worker. This removes a lot of overhead.

Of course that's easier said than done, so the good news is I spend a couple of hours on this. Tests pass, locally at least… let me know what you think 😄 

There was some bike shedding as well, mostly in the handles and reference counting.

**Comment by dnlup** [comment]
> @dnlup so I didn't end up looking into that specific error, but at the overall approach. I think we got stuck in making and discussing local changes. Now that shared workers only work in threads we can actually allocate the message ports in the test worker and forward them to the shared worker. This removes a lot of overhead.
> 
> Of course that's easier said than done, so the good news is I spend a couple of hours on this. Tests pass, locally at least… let me know what you think 
> 
> There was some bike shedding as well, mostly in the handles and reference counting.

I like this approach! I admit I was a little hesitant about removing some parts because I was afraid they were important pieces. This is much cleaner now.

I see the error popping out now on the CI:

* https://github.com/avajs/ava/pull/2706/checks?check_run_id=2254831559#step:7:4350

I still have it locally too. I am using:

* Ubuntu 20.04.2
* Node 14.16.0

I have tried to debug it (as I tried before), and it looks like the error from the test worker is being passed correctly. I still haven't understood why `ava` errors with `Promise` never resolved.

I apologize that this is taking so long, and thank you for the help.

**Comment by novemberborn** [comment]
Aha! I can reproduce this locally as well. Turns out a `beforeExit` event was triggered, even though the test was still waiting for the crash error to propagate.

The reference management didn't quite seem right, but even then this seems to require a workaround. So now I'm keeping a timer referenced for up to a second while we wait for Node.js to respect us referencing the message port.



> I apologize that this is taking so long

Not at all! I don't think either of us are getting paid for this, so it's healthier to treat it as a hobby… and those require spare time 😄

**Comment by dnlup** [comment]
> The reference management didn't quite seem right, but even then this seems to require a workaround. So now I'm keeping a timer referenced for up to a second while we wait for Node.js to respect us referencing the message port.

So the issue was `ref`/`unref` not working properly in Node itself?

**Comment by novemberborn** [comment]
> So the issue was `ref`/`unref` not working properly in Node itself?

It would seem like it. Referencing the message port did not stop the thread from exiting (test files now run in worker threads in AVA 4). Sometimes though the test passed, so there's probably a race condition somewhere. Referencing the timer seems to be more stable.

**Pull request merged**